### PR TITLE
Google Preview Link

### LINKED ISSUE #2178: Display "snippets" from Google Books if possible
Displaying Google snippets of non-public items will help users evaluate items, and reduce unnecessary pickup or scanning requests.

This is only for materials that we don't have a better alternative for (eg. Hathi, already-completed CDL).
## Additional implementation details:

* UPDATE 9/10/2020 - Implement this feature using the link out to Google Book Search (example: https://developers-dot-devsite-v2-prod.appspot.com/books/examples/previewwizard-linktogbs), not the embed. 
* Display disclaimer text stating that this is a preview, and that PUL does not recommend/endorse purchasing options on the following page ([text for this](https://docs.google.com/document/d/1xbjEzo4p0vcblY9-KE6Fsv73kNHj151iJysYMCSJnvc/edit)). Initially, display text in line with the link itself. If the text is too long, use a click-through modal that is presented only once per session.
* If possible, link out to the New Google Books view of the page (de-emphasizes purchase). Example: [old view](https://books.google.com/books?id=VpNa9UckT24C&printsec=frontcover&source=gbs_ViewAPI&hl=en#v=onepage&q&f=false) | [new view (preferred)](https://www.google.com/books/edition/_/VpNa9UckT24C?hl=en&gbpv=1&pg=PP1)
* Add the snippet link to the "Show" page in the catalog, where the UV display for items with digital representations
* Make it clear through labeling the link that this is a snippet, not the full book (based on user feedback from a previous implementation)
* Only attempt to match ISBN numbers
* Only display this for items that are print monographs (no serials, no e-resources)
* ~~Link the thumbnail image in the "Show" page to the snippet viewer at the bottom of the page, if possible~~

### PR DESCRIPTION
**Pull request by tpendragon**
Closes #2178

### CODE REVIEW

File: spec/javascript/orangelight/insert_online_link.spec.js [hunk 1]
```diff
@@ -33,4 +33,15 @@ describe('insert_online_link', function() {
     expect(anchor.getAttribute("href")).toEqual("#view")
     expect(anchor.getAttribute("target")).toEqual("_self")
   })
+
+  test("insert_online_link() can accept a content builder", () => {
```
Comment (hackmastera) [inline spec/javascript/orangelight/insert_online_link.spec.js:37]: 👍

File: app/javascript/orangelight/google_books_snippets.es6 [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  process_google_response(response) {
+    for(const key in response) {
+      const result = response[key]
+      if(result.preview === "partial" || result.preview === "full")
+      {
+        const url = new URL(result.preview_url)
+        const bookId = url.searchParams.get("id")
+        const link = `https://www.google.com/books/edition/_/${bookId}?hl=en&gbpv=1&pg=PP1`
+        const content = (link, target) => {
+          return `<a href="${link}" target="${target}">Google Preview</a><p>${this.contentStatement}</p>`
```
Comment (hackmastera) [inline app/javascript/orangelight/google_books_snippets.es6:35]: Specify "partial" or "full" in the link text?

**Review by hackmastera — CHANGES_REQUESTED** [review]
I think it would be nice to indicate whether it is a full or partial preview.

Also I think the requirement to exclude serials and journals is not yet addressed. Is this desired before the feature is deployed or should we create a follow-up issue? @mzelesky's comment on the previous PR: https://github.com/pulibrary/orangelight/pull/2225#issuecomment-685734042

File: app/javascript/orangelight/google_books_snippets.es6 (hunk 2 above)
Comment (tpendragon) [reply app/javascript/orangelight/google_books_snippets.es6:35]: If it's full is it even a preview anymore? Possible names fluttering in my head are `Google Books (Partial)`, `Google Preview (Partial)`, `Google Book Viewer (Partial)`, `Google Book Viewer`. Maybe input from @escowles or @jpstroop ?

Comment (jpstroop) [reply app/javascript/orangelight/google_books_snippets.es6:35]: I thought we we scoping to previews only and that the full views would be covered by Hathi, but you're right, there are probably exceptions. `Google Books (Partial View)` and `Google Books (Full View)` makes sense to me. I'll look at the disclaimer text too and see if that needs adjusting.

Comment (jpstroop) [reply app/javascript/orangelight/google_books_snippets.es6:40]: Disclaimer updated once again: https://docs.google.com/document/d/1xbjEzo4p0vcblY9-KE6Fsv73kNHj151iJysYMCSJnvc/edit

Comment (tpendragon) [reply app/javascript/orangelight/google_books_snippets.es6]: Screen Shot 2020-09-16 at 10 08 22 AM

**Comment by tpendragon** [comment]
I added not running this code if the item's a Journal. I'm not really sure how to do e-resources only. @mzelesky do you have an example on hand of one in the catalog? I need something in the DOM to key off of, ideally.

**Comment by tpendragon** [comment]
@hackmastera Okay I think this is finally ready.

File: spec/requests/request_spec.rb [hunk 3]
```diff
@@ -347,6 +347,16 @@
     end
   end
 
+  describe 'location metadata' do
```
Comment (hackmastera) [inline spec/requests/request_spec.rb:350]: I don't think I'd expect to find this test in this file. How did you decide to put it here?

**Review by hackmastera — CHANGES_REQUESTED** [review]
I just have the one question inline

File: spec/requests/request_spec.rb (hunk 3 above)
Comment (tpendragon) [reply spec/requests/request_spec.rb:350]: This is where the test for the identifiers was - seemed like where the catalog show specs are.

**Review by hackmastera — APPROVED** [review]

**Pull request merged**